[P095] Add support for extra TFT displays (ILI934x/ILI948x)

tonhuisman opened the pull request:
On request via multiple channels ([Forum req 1](https://www.letscontrolit.com/forum/viewtopic.php?f=5&t=8836), [forum req 2](https://www.letscontrolit.com/forum/viewtopic.php?f=5&t=8913&p=56456), PM) I have done some investigation and found that the displays mentioned here all are quite similar, except for the initialization and rotation support. As the initialization for the Adafruit_ILI9341 library is very structured and can easily be extended, I've added these display models to the base support for ILI9341 (that also should work with ILI9340):
- ILI9342 240x320px (as used on some M5Stack units)
- ILI9481 320x480px (f.e. used on this [LilyGo Lily PI unit](http://www.lilygo.cn/prod_view.aspx?TypeId=50032&Id=1353))
  - Has support for some specific hardware variations: CPT29, PVI35, AUO317, CMO35, RGB, CMI7 and CMI8.

Skipped for now, because these type only seem to be available in 8 bit parallel (DMA) interface, not in regular SPI:
- ILI9486 320x480px
- ILI9488 320x480px

This includes a modified version of the Adafruit_ILI9341 library, where I added the specific Initialization and Rotation handling.
Adapted initialization/rotation is based on the famous [Bodmer TFT_eSPI](https://github.com/Bodmer/TFT_eSPI) library, that supports many displays, but isn't very well suited to include in ESPEasy, because of the conditional compilation required.

NB: Original development was done in combination with P116 and the AdaGFX_Helper module, but I created a separate PR to improve testability, so those parts are not included here.
NB2: Touch support is usually by an XPT2046 driver, that is already supported by ESPEasy via plugin P099, other touch drivers is being worked on, like the FT62x6 in planned plugin P123. Also, support for GT911/GT9110 touch driver has been planned, no plugin ID reserved yet.

TODO:
- [x] Update documentation ~(may be postponed until the PR for plugin P116 is merged, as there I've already created documentation for this plugin)~
- [x] Testing: Help wanted 🌞 (as usual, I don't own most of the hardware, so I can't test myself)
  - [x] ILI9341 ~(to avoid regression errors, tested locally already, confirmation requested)~
  - [x] ILI9342 (tested on M5Stack Core)
  - [x] ILI9481 (base-type) (tested on LilyGo Lily Pi, exception: changing rotation _after_ displaying anything causes distorted display, like moving, rotating and/or mirroring existing content)
    - [x] ILI9481 sub-types, (don't know how to distinguish the separate types, though) (also tested on Lily Pi)
  - ILI9486 (~if available~Seems to only be available in 8 bit parallel versions, not in regular SPI, ignored for now)
  - ILI9488 (~if available~Seems to only be available in 8 bit parallel versions, not in regular SPI, ignored for now)
- [x] Fixed: Overwriting text when non-default font was used (may cause some minor flashing, as the space is cleared and re-painted)
- [x] Added: Feature to display text centered within a specific width (using `tft,txtfull...` command)
- [x] Added: Enable SD-card support, as AdafruitGFX_helper can show .bmp files from SD storage, and some displays have an SD slot
- [x] Make ILI9486/ILI9488 a compile-time option to save a little space.

When testing, please pay attention to the Rotation setting & command (`tft,rot,<0..3>`), they should work as advertised, or else a bug crept in 😮 (exception: ILI9481, see above)

Resolves #4060

tonhuisman committed:
[P095] Add support for extra TFT displays (ILI934x/ILI948x)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Reformat sources using Uncrustify

tonhuisman commented:
The selector for the display type: (Updated: 2022-07-17)

![P095_TFTDisplayModelOptions](https://user-images.githubusercontent.com/1922393/179405835-bf80b718-90c6-42c7-8c4d-da845d9d127a.png)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] String allocation improvements

TD-er commented on src/_P095_ILI9341.ino:
```diff
@@ -557,19 +557,21 @@ boolean Plugin_095(uint8_t function, struct EventStruct *event, String& string)
+          log += F("Parsed command = \"");
+          log += string;
+          log += '"';
+          addLog(LOG_LEVEL_INFO, log);
```
Why not an `addLogMove` ?

tonhuisman replied:
Missed that when handling the merge 😒

tonhuisman committed:
[P095] Restore accidently lost addLogMove

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Fix merge issue

TD-er commented on platformio_esp82xx_base.ini:
```diff
@@ -49,7 +49,7 @@ board_build.f_cpu         = 80000000L
 build_unflags             = -DDEBUG_ESP_PORT
                             -fexceptions
-lib_deps                  = td-er/ESPeasySerial @ 2.0.8, I2Cdevlib-Core, Adafruit ILI9341, Adafruit GFX Library, LOLIN_EPD, adafruit/Adafruit BusIO @ ^1.10.0, bblanchon/ArduinoJson @ ^6.17.2, VL53L0X @ 1.3.0, SparkFun VL53L1X 4m Laser Distance Sensor @ 1.2.9, td-er/RABurton ESP8266 Mutex @ ^1.0.2, td-er/SparkFun MAX1704x Fuel Gauge Arduino Library @ ^1.0.1, ESP8266HTTPUpdateServer, FrogmoreScd30, ...
+lib_deps                  = td-er/ESPeasySerial @ 2.0.8, I2Cdevlib-Core, ESP8266WebServer, Adafruit, Adafruit GFX Library, LOLIN_EPD, adafruit/Adafruit BusIO @ ^1.10.0, bblanchon/ArduinoJson @ ^6.17.2, VL53L0X @ 1.3.0, SparkFun VL53L1X 4m Laser Distance Sensor @ 1.2.9, td-er/RABurton ESP8266 Mutex @ ^1.0.2, td-er/SparkFun MAX1704x Fuel Gauge Arduino Library @ ^1.0.1, ESP8266HTTPUpdateServer, Frog ...
```
I doubt there is a library called just "Adafruit"

tonhuisman replied:
That lib is found by its library name 'Adafruit ILI9341', that I pulled local to ESPEasy in this PR to add support for the extra display models. Maybe I should have changed the name, but that is still an option.
In the previous merge of `mega` I messed up this name, but it has been corrected in the last commit.

TD-er replied:
OK, strange, as it is the worst possible name for a library.

tonhuisman committed:
[P095] Make enum values explicit

TD-er committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

TD-er commented on platformio_esp32_envs.ini:
```diff
@@ -11,7 +11,7 @@ lib_ignore                = ESP8266WiFi, ESP8266Ping, ESP8266WebServer, ESP8266H
 [esp32_common]
 extends                   = common, core_esp32_3_3_2
 platform                  = ${core_esp32_3_3_2.platform}
-lib_deps                  = td-er/ESPeasySerial @ 2.0.8, I2Cdevlib-Core, adafruit/Adafruit ILI9341 @ ^1.5.6, Adafruit GFX Library, LOLIN_EPD, adafruit/Adafruit BusIO @ ^1.10.0, VL53L0X @ 1.3.0, SparkFun VL53L1X 4m Laser Distance Sensor @ 1.2.9, td-er/SparkFun MAX1704x Fuel Gauge Arduino Library @ ^1.0.1, ArduinoOTA, ESP32HTTPUpdateServer, FrogmoreScd30, Multi Channel Relay Arduino Library, SparkF ...
```
Why remove the version nr and the adafruit prefix?

tonhuisman replied:
I pulled the library local to make changes to the initialization, as is the easiest way to add support for the extra display models. Don't expect that a) Adafruit is interested in these mods, and b) seeing their often delayed response to PR's, it would take considerable time before being available for general use, once accepted.

TD-er replied:
Maybe we should relabel it by making our own version/fork of it.
This may lead to some random choice by PlatformIO for which version will be picked.

tonhuisman replied:
Renamed library to have an 'ESPEasy' postfix.

TD-er commented on src/src/PluginStructs/P095_data_struct.h:
```diff
@@ -38,60 +75,92 @@ enum class ILI9xxx_type_e : uint8_t {
+  String              _commandTrigger;
+  uint16_t            _fgcolor;
+  uint16_t            _bgcolor;
+  bool                _textBackFill;
```
Shouldn't these be properly initialized to some value?

tonhuisman replied:
The constructor takes care of that.

TD-er replied:
By initializing in the .h when declaring the variables, you will be sure it is always set, no matter which constructor is being used.
And unless you declare other constructors explicitly with `= delete;` you can't be sure those will not be used by accident. (e.g. assignment constructor)

tonhuisman committed:
[P095] Rename local copy of Adafruit ILI9341 library, initializations

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Updated documentation

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'feature/P095-add-extra-display-types' of https://github.com/tonhuisman/ESPEasy-1 into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

TD-er committed:
[PlatformIO] Do not extend multiple PIO envs

TD-er committed:
[PlatformIO] Change src_filter into build_src_filter for PIO 6.0.0

See: https://github.com/platformio/platformio-core/issues/4270

TD-er committed:
[PlatformIO] Revert to PlatformIO 5.2.5

TD-er committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Make Splash at start configurable, docs updated

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Fix for displaying text on previous location was not overwriting old content
Add Text print mode for centered text
Extend txtfull subcommand to enable centered text

tonhuisman committed:
[P095] Update documentation

tonhuisman committed:
[P095] Implement splash without delay()

(cherry picked from commit e897a337753e3b3aecadc0bcfe53e1b10d3080a1)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Use ESPEasy-specific ILI9341 extended library

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[Build] Enable SD-card support for Display and NeoPixel sets

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[ILI9341 lib] Extend supported ILI9481 sub-types (CMI7 and CMI8, borrowed from TFT_eSPI)

tonhuisman committed:
[P095] Extend supported ILI9481 display models

tonhuisman committed:
[P095] Update documentation

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
[P128] Fix compiler warning

(cherry picked from commit acd3c5073fc289acbb0adf9a65e1f1cb13aeb0d0)

tonhuisman committed:
[SDcard] Remove unused variable in handle_SDfilelist()

(cherry picked from commit bc5d83996e3a5397c662c80ebfab773a678ee615)

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
[Build] Leave SD-card support out for any NEOPIXEL ESP8266 build to make it fit again

tonhuisman committed:
[P095] Disable unavailable displays (ILI9486/ILI9488)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'feature/P095-add-extra-display-types' of https://github.com/tonhuisman/ESPEasy-1 into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'feature/P095-add-extra-display-types' of https://github.com/tonhuisman/ESPEasy-1 into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
Merge branch 'mega' into feature/P095-add-extra-display-types

tonhuisman committed:
[P095] Optimize code and reduce string use to make build fit
